### PR TITLE
BF runs scripts fixed, majority EltWise kernels pass 

### PR DESCRIPTION
**Pull request by nkmellem**

### CODE REVIEW

**Review by alheinecke — CHANGES_REQUESTED** [review]
Can you please modify the test scripts such that 4 4 4 is only tested ones. I guess the best way to do this to keep your old loops (1 2 4 for in and out) and just skip in case of 1/2 2/1 pairs for in out.

I'm not sure about the replicate cols change. That would indicate there is bug in master, but it's passing on master.

**Comment by alheinecke** [comment]
any update on this?

File: src/generator_mateltwise_misc_avx_avx512.c [hunk 1]
```diff
@@ -251,7 +251,7 @@ void libxsmm_generator_replicate_col_var_avx_avx512_microkernel( libxsmm_generat
       i_micro_kernel_config->alu_mov_instruction,
       i_gp_reg_mapping->gp_reg_param_struct,
       LIBXSMM_X86_GP_REG_UNDEF, 0,
-      56,
```
[the remark this answers is not in the document]
Comment (alheinecke) [reply src/generator_mateltwise_misc_avx_avx512.c]: we resolved that.

File: src/generator_common_x86.c [hunk 2]
```diff
@@ -418,7 +418,7 @@ void libxsmm_x86_instruction_vpermd_16way_avx2( libxsmm_generated_code*
             i_vec_tmp1,
             i_vec_tmp0,
             i_vec_result,
-            0, 0, 0, (unsigned short)(i_vec_result) << 4);
+            0, 0, 0, (unsigned short)((i_vec_result) << 4));
```
Comment (alheinecke) [inline src/generator_common_x86.c:421]: What issue is this fixing?

File: samples/eltwise/kernel_test/unary_replicate_col_var_mixed_eqld.sh [hunk 3]
```diff
@@ -44,6 +44,19 @@ do
       done
     done
   done
+#  for PREC_IN in 1 4
```
Comment (alheinecke) [inline samples/eltwise/kernel_test/unary_replicate_col_var_mixed_eqld.sh:47]: Why comment, can you please clean-up

File: samples/eltwise/kernel_test/unary_dropout_mixed_gtld.sh [hunk 4]
```diff
@@ -29,15 +29,24 @@ do
...
   do
-    for PREC_OUT in 1 2 4
+    for PREC_OUT in 2 4
     do
       ./eltwise_unary_dropout F 0 ${PREC_IN} ${PREC_OUT} ${M} ${N} 100 100
       ./eltwise_unary_dropout F 1 ${PREC_IN} ${PREC_OUT} ${M} ${N} 100 100
       ./eltwise_unary_dropout B 1 ${PREC_IN} ${PREC_OUT} ${M} ${N} 100 100
     done
   done
+#  for PREC_IN in 1 4
```
Comment (alheinecke) [inline samples/eltwise/kernel_test/unary_dropout_mixed_gtld.sh:41]: Why comment, can you please clean-up? It seems to disabled BF8 test... what's the fix?

File: samples/eltwise/kernel_test/unary_dropout_mixed_eqld.sh [hunk 5]
```diff
@@ -29,15 +29,24 @@ do
...
   do
-    for PREC_OUT in 1 2 4
+    for PREC_OUT in 2 4
     do
       ./eltwise_unary_dropout F 0 ${PREC_IN} ${PREC_OUT} ${M} ${N} ${LDI} ${LDI}
       ./eltwise_unary_dropout F 1 ${PREC_IN} ${PREC_OUT} ${M} ${N} ${LDI} ${LDI}
       ./eltwise_unary_dropout B 1 ${PREC_IN} ${PREC_OUT} ${M} ${N} ${LDI} ${LDI}
     done
   done
+#  for PREC_IN in 1 4
```
Comment (alheinecke) [inline samples/eltwise/kernel_test/unary_dropout_mixed_eqld.sh:41]: This seems to disable BF8, what is the issue?

**Review by alheinecke — CHANGES_REQUESTED** [review]
Can please clean-up the scripts as discussed in the previous reviews, it also seems that some kernels (like dropout don't test BF8)

File: samples/eltwise/kernel_test/unary_dropout_mixed_gtld.sh (hunk 4 above)
Comment (nkmellem) [reply samples/eltwise/kernel_test/unary_dropout_mixed_gtld.sh]: resolved

File: samples/eltwise/kernel_test/unary_replicate_col_var_mixed_eqld.sh (hunk 3 above)
Comment (nkmellem) [reply samples/eltwise/kernel_test/unary_replicate_col_var_mixed_eqld.sh]: cleaned up. combined the tests.

File: src/generator_common_x86.c (hunk 2 above)
Comment (nkmellem) [reply src/generator_common_x86.c:421]: apparently none.  you can revert back to the old code.

**Comment by nkmellem** [comment]
There is still a memory overwrite issue in the BF8 path. I could not yet locate the offending code. I have added temporary padding to buffers (4096 alignment) to fix the relicate_col test. ReLU and X2 are still failing for the same reasons.

**Comment by alheinecke** [comment]
I suggest debugging using the X^2 case as this is a very simple kernel, if this fails everything should fail.

File: samples/eltwise/eltwise_unary_simple.c [hunk 6]
```diff
@@ -306,9 +306,9 @@ int test_unary_op( const libxsmm_blasint M, const libxsmm_blasint N, const libxs
 
   libxsmm_rng_set_seed(1);
 
-  in        = (char*) libxsmm_aligned_malloc( LIBXSMM_TYPESIZE(dtype_in) *N*LIBXSMM_MAX(M,ldi), 64 );
-  out       = (char*) libxsmm_aligned_malloc( LIBXSMM_TYPESIZE(dtype_out)*N*ldo,                64 );
-  out_gold  = (char*) libxsmm_aligned_malloc( LIBXSMM_TYPESIZE(dtype_out)*N*ldo,                64 );
+  in        = (char*) libxsmm_aligned_malloc( LIBXSMM_TYPESIZE(dtype_in) *N*LIBXSMM_MAX(M,ldi), 4096 );
```
Comment (alheinecke) [inline samples/eltwise/eltwise_unary_simple.c:309]: as mentioned in my outer comment, starting debugging with X^2 should be the easiest.

File: samples/eltwise/kernel_test/unary_copy_mixed_eqld.sh [hunk 7]
```diff
@@ -31,11 +31,24 @@ do
   LDI=`echo ${i} | awk -F"_" '{print $3}'`
   LDO=`echo ${i} | awk -F"_" '{print $4}'`
   echo ${M} ${N} ${LDI} ${LDI}
-  for PREC_IN in 1 2 4
+  for PREC_IN in 2 4
   do
     for PREC_COMP in 4
     do
-      for PREC_OUT in 1 2 4
+      for PREC_OUT in 2 4
```
[the remark this answers is not in the document]
Comment (alheinecke) [reply samples/eltwise/kernel_test/unary_copy_mixed_eqld.sh]: resolved in last commit

File: samples/eltwise/eltwise_unary_simple.c [hunk 8]
```diff
@@ -306,8 +306,11 @@ int test_unary_op( const libxsmm_blasint M, const libxsmm_blasint N, const libxs
 
   libxsmm_rng_set_seed(1);
 
+  int out_size = LIBXSMM_TYPESIZE(dtype_out)*N*ldo;
+  out_size = out_size + (64 - (out_size%64));
```
Comment (alheinecke) [inline samples/eltwise/eltwise_unary_simple.c:310]: This indicates a buffer overrun bug in the kernel.

File: src/generator_mateltwise_misc_avx_avx512.c [hunk 9]
```diff
@@ -359,7 +359,7 @@ void libxsmm_generator_replicate_col_var_avx_avx512_microkernel( libxsmm_generat
   /* In this case we have to use CPX replacement sequence for downconverts... */
   if ((io_generated_code->arch < LIBXSMM_X86_AVX512_CPX) && (io_generated_code->arch != LIBXSMM_X86_AVX512_VL256_CPX) && (LIBXSMM_DATATYPE_BF16 != LIBXSMM_GETENUM_INP( i_mateltwise_desc->datatype )) && (LIBXSMM_DATATYPE_BF16 == LIBXSMM_GETENUM_OUT( i_mateltwise_desc->datatype )) ) {
     libxsmm_generator_vcvtneps2bf16_avx512_clean_stack( io_generated_code, LIBXSMM_X86_GP_REG_RAX );
-  } else if (LIBXSMM_DATATYPE_BF8 == LIBXSMM_GETENUM_OUT( i_mateltwise_desc->datatype ) ) {
+  } else if ( (LIBXSMM_DATATYPE_BF8 != LIBXSMM_GETENUM_INP( i_mateltwise_desc->datatype )) && (LIBXSMM_DATATYPE_BF8 == LIBXSMM_GETENUM_OUT( i_mateltwise_desc->datatype )) ) {
```
Comment (alheinecke) [inline src/generator_mateltwise_misc_avx_avx512.c:362]: Why != and == a couple of lines above?

File: samples/eltwise/eltwise_unary_dropout.c [hunk 10]
```diff
@@ -323,6 +323,10 @@ int test_dropout_fwd( const libxsmm_blasint bitm, const libxsmm_blasint M, const
     if ( norms_out.normf_rel > 0.00001 ) {
       ret = EXIT_FAILURE;
     }
+  } else if ( (dtype_in == LIBXSMM_DATATYPE_BF8) && (dtype_out == LIBXSMM_DATATYPE_BF8) && (dtype_comp == LIBXSMM_DATATYPE_F32) ) {
+    if ( norms_out.normf_rel > 0.125 ) {
+      ret = EXIT_FAILURE;
```
Comment (alheinecke) [inline samples/eltwise/eltwise_unary_dropout.c:328]: In unary simple the error bound is 0.25, why here 0.125?

**Review by alheinecke — APPROVED** [review]

**Pull request merged**